Add option to show full date and select date format

hjiangsu opened the pull request:
## Pull Request Description

This PR adds the ability to show full dates on posts, and select a date format from a pre-selected list. In addition, I added a way to add subtitles to a given PickerItem

<!--- Please describe what was changed -->

## Issue Being Fixed

<!-- Please describe the problem that is being fixed and, if applicable, reference a GitHub issue -->

Issue Number: #1114

## Screenshots / Recordings


https://github.com/thunder-app/thunder/assets/30667958/6633a69a-961a-463a-9fd5-ce19dc219ca9



<!-- This section is optional but highly recommended to show off your changes! -->

## Checklist

- [ ] Did you update CHANGELOG.md?
- [ ] Did you use localized strings where applicable?
- [ ] Did you add `semanticLabel`s where applicable for accessibility?

K4LCIFER commented:
Looks awesome! I would maybe make the date more of a grey color so that it isn't as jarring. Also would it be possible for the user to specify their own date format? It seems unnecessary to try and specify every possible date combination that a user may want. I would use one as a sane default (probably something of the form "Mar 7, 2024 12:23") and then allow the user to specify alternatives if they wish.

micahmo commented on lib/settings/pages/post_appearance_settings_page.dart:
```diff
@@ -511,6 +540,35 @@ class _PostAppearanceSettingsPageState extends State<PostAppearanceSettingsPage>
+            child: ListOption(
+              description: l10n.dateFormat,
+              disabled: !showFullPostDate,
+              value: ListPickerItem(
```
For some reason the initially selected value is not shown as highlighted in the list the first time you open the modal. Once you pick a value, subsequently it is highlighted. So I think you'd have to clear your preferences to repro. But it's visible in your demo.

Maybe it's because the `DEFAULT_DATE_FORMAT` constant is not a reference match for any of the entries in the `dateFormats` list.

micahmo reviewed (approved):
This is super cool!

The only thing I could possibly suggest is to have a "system" setting where the date format follows the locale of the device. This is pretty common in date formatting, and it looks like it is possible in Dart by constructing a `DateFormat()` without any parameters.

**EDIT**: I take it back. Constructing a `DateFormat()` with no parameters always falls back to a hard-coded format, `yMMMMd` plus `jms`.

https://github.com/dart-lang/i18n/blob/c6c6c25030c88ca4db7cacc6ad312c8f3d1ba357/pkgs/intl/lib/src/intl/date_format.dart#L724-L727

I will see if there's another way to construct a local format.

**EDIT2:** Ok, sorry for all the edits! We can use the existing `yMMMMd` and `jms` formats while passing the current locale.

``` dart
  final DateTime now = DateTime.now();
  
  print(Intl.systemLocale); // Prints: en_US
  DateFormat dateFormat = DateFormat.yMMMMd(Intl.systemLocale);
  DateFormat timeFormat = DateFormat.jms(Intl.systemLocale);
  print('${dateFormat.format(now)} ${timeFormat.format(now)}'); // Prints: March 7, 2024 3:59:22 PM
```
---
To test that it really does respect the locale...
``` dart
final DateTime now = DateTime.now();
  
  initializeDateFormatting('en_GB');
  DateFormat dateFormat = DateFormat.yMMMMd('en_GB');
  DateFormat timeFormat = DateFormat.jms('en_GB');
  print('${dateFormat.format(now)} ${timeFormat.format(now)}'); // Prints: 7 March 2024 16:04:04
```

I completely agree with @K4LCIFER on having a "sane default", but unfortunately that means different things to different people around the world. 😊

hjiangsu commented:
> I would maybe make the date more of a grey color so that it isn't as jarring

I can maybe adjust the date colour if using custom dates, otherwise it'll default to what it is now!

> would it be possible for the user to specify their own date format?

This was my plan for a future enhancement! The reason why I opted to show a few pre-selected options for now is to reduce complexity of the feature. To allow for user-specific date formats, there would be some additional requirement to validate the user input, or implement it in a way that's intuitive to use (e.g., maybe something similar to the post card metadata customization with chips for different sections)

hjiangsu commented:
> The only thing I could possibly suggest is to have a "system" setting where the date format follows the locale of the device.

I like this suggestion! I'll change it so that it does that based on what you mentioned with addling the system locale!

micahmo commented:
> I'll change it so that it does that based on what you mentioned with addling the system locale!

Sorry for commenting so much on this one. One more thought to add is that, if the system format happens to be a duplicate one of the predefined formats, you could had that so it's only displayed once. You can see what the system format would be by checking `pattern`.

``` dart
'${dateFormat.pattern} ${timeFormat.pattern}'
```

hjiangsu commented:
Alright, I made some slight changes based on the feedback!
- When the full date is shown, it shows up as slightly dimmer on the feed page
- By default, the full date will use the system's locale to determine the format of the date
- Removed any duplicated formats if it exists (https://github.com/thunder-app/thunder/pull/1174/commits/53184a57e70836ad1f3b3ded8ee38a70fe59ef4d)
- I also removed the seconds from the date. My reason for this is that seconds don't contribute as much information as the rest of the timestamp, but takes up horizontal real-estate which makes it feel "clunky"

Here's an updated video with the changes. Let me know what you think!


https://github.com/thunder-app/thunder/assets/30667958/b4324418-91e3-4eab-bcc5-8ddba5242219

micahmo commented on lib/settings/pages/post_appearance_settings_page.dart:
```diff
@@ -79,6 +80,15 @@ class _PostAppearanceSettingsPageState extends State<PostAppearanceSettingsPage>
+  DateFormat? selectedDateFormat;
+
+  /// List of available date formats to select from
+  List<DateFormat> dateFormats = [DateFormat.yMMMMd(Intl.systemLocale).add_jm()];
```
Is there any concern that `add_jm()` is not taking a locale here? You can call `DateFormat.jm(<locale>)`, but I couldn't find a great way to combine to formats that both take a locale.

hjiangsu replied:
I _believe_ it takes the locale from the date format? `DateFormat.yMMMMd(<locale>)` and `DateFormat.jm(<locale>)` are just constructors which create a DateFormat object with the applied locale.

`add_jm()` is then just a member function which changes the internal DateFormat pattern. When we finally parse the date format, I think that's when the locale is applied. Also, it seems like by default, it uses `Intl.systemLocale` for the locale if none is passed in.

https://api.flutter.dev/flutter/intl/DateFormat/DateFormat.d.html

micahmo replied:
Sweet, thanks for confirming that!

micahmo reviewed (approved):
Awesome feature and nice improvement from feedback!
